docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -735,28 +563,86 @@ As database IDs are all globally unique, it's fine to replace table ID with data
+
+CockroachDB does not look to have something directly comparable to `PLACEMENT POLICY`, but it does have the ability to specify "replication zones" for "system ranges" such as default, meta, liveness, system, timeseries. Before dropping `ALTER TABLE t1 ADD PLACEMENT` from this proposal, it was investigated the CockroachDB does not support this syntax, presumably for simplification and minimising similar risks of misconfiguration.
+
+## Unresolved Questions
+
+### Compliance Requirements
+
+For compliance use-cases, it is clear that data at rest should reside within a geographic region. What is not clear is which (if any) circumstances data in transit is permitted to leave. There are several known issues which will need to be resolved:
 
-## Ecosystem tools
+* **Backup**: The `BACKUP` SQL command (and br) accept a single location such as `s3://bucket-location/my-backup` to centralize the backup. This centralization likely violates compliance requirements (assuming there are >=2 requirements that conflict on a single cluster). Backing up segments of data individually is both an inconsistent backup, and likely to result in misconfiguration which violates compliance rules. Users have a reasonable expectation of backup integration with compliance placement rules.
+* **CDC**: Similar to `BACKUP`, it should not be possible to subscribe to changes for data in a different jurisdiction.
+* **DDL**: The current implementation of DDL uses a centralized DDL owner which reads data from relevant tables and performs operations such as building indexes. The _read_ operation may violate compliance rules.
+* **Internal SQL**: Similar to DDL, several centralized background tasks, such as updating histograms/statistics need to be able to _read_ the data.
+* **DML**: It is not yet known which restrictions need to be placed on user queries. For example, if a poorly written user-query does not clearly target the `USA` partition when reading user data, should it generate an error because the `EUROPE` partition needs to be read in order for the semantics of the query to be correct? This may cause problems in development and integration environments if the restrictions can not be distilled into environments with smaller topologies.
 
-Many tools are based on binlog or metadata. For example, TiDB-binlog is based on binlog, while Lightning and Dumpling are based on metadata. Placement rules need to be compatible with these tools.
+### Behaviors
 
-If the downstream is not TiDB, no change needs to be made. But even if it is TiDB, TiKV nodes may have a different geographical topology, which means the labels of TiKV nodes may be different. In this case, placement rules can not be enforced on them.
+* The proposal says that `DROP PLACEMENT POLICY` should error if the policy is in use by any objects. This should be discussed, because it is not a common behavior in MySQL and may create some annoying debugging scenarios?
```

Comment:
I would vote not to error. From the implementation perspective, policies are duplicated anyway, you don't know if it is in use trivially. And we don't have `EDIT PLACEMENT POLICY`, why bother to track the using state?

Reply:
What is the behavior of named placement policies if the named placement is dropped? Does `SHOW CREATE TABLE` just show the `CONSTRAINTS` without the `POLICY=x`? That seems acceptable to me. Showing the name (which no longer exists) does not.

Reply:
Yes. Actually, it becomes easier if we don't output `POLICY=x` at all... :joy:

Reply:
I think we should output it though. If _table_options_ on `CREATE TABLE` + `ALTER TABLE` are not orthogonal, it creates strange debugging issues where the altered options don't seem to "stick" (at least for me, based on my expectations as a MySQL user).

IMHO there are two valid behaviors (if DROP is not an error):
1. The `DROP` "cascades", and all tables that have this policy are now restored to having the default placement.
2. The `DROP` just removes the named policy, and the output will instead show the same options but as `PLACEMENT CONSTRAINTS ..`

The closest example I can think of in MySQL is dropping a ROLE which is assigned to a set of users. In which case, the [MySQL behavior](https://dev.mysql.com/doc/refman/8.0/en/drop-role.html) here matches (1):

> A dropped role is automatically revoked from any user account (or role) to which the role was granted. Within any current session for such an account, its adjusted privileges apply beginning with the next statement executed.

Reply:
> The DROP "cascades", and all tables that have this policy are now restored to having the default placement.

I would vote for this. It makes reasoning easier.

> The DROP just removes the named policy, and the output will instead show the same options but as PLACEMENT CONSTRAINTS ..

Maybe you will add a policy with the same name, then it becomes a trouble.

Reply:
I prefer (1) too, but it has another consequence: the other rule changes are currently defined as synchronously applied, but `DROP POLICY` waiting for all tables to change to the default scheme may be weird and confusing.

It may be better to define placement changes as asyncronous, and can be monitored with `SHOW PLACEMENT`. I will add this item to the "unanswered questions" section.

Reply:
It is async anyway. PD HTTP API only promise it is posted and will take effect eventually. You could refer the discussion https://github.com/pingcap/tidb/pull/26221#discussion_r669388467.

Reply:
The behavior of DROP is now documented. I have 'async' captured in the unresolved questions, so I am going to resolve this comment for now.